### PR TITLE
Rewrite of regex interpolation section

### LINKED ISSUE #2599: Regex interpolation
## The problem

[Regex interpolation](https://docs.perl6.org/language/regexes#Regex_interpolation) documentation is too laconic:

> There are four ways you can interpolate a string into regex as a pattern. That is using `$pattern`, `$($pattern)`, `<$pattern>` or `<{$pattern.method}>`.

Ok, it very clear and understandable.

> Note that the two examples above interpolate the string lexically, while `<$pattern>` and `<{$pattern.method}>` causes implicit `EVAL`, which is a known trap.

However, the second paragraph is not. What does "interpolate the string lexically" mean? (I just re-read section "[Interpolation: qq](https://docs.perl6.org/language/quoting#Interpolation:_qq)" and did not find the term "lexical interpolation".) What the difference between `$pattern` and `$($pattern)`? Ok, "`<$pattern>` and `<{$pattern.method}>` causes implicit `EVAL`", but what does it _really_ mean? Will the result of interpolation interpreted as regex (`/$pattern/` in Perl5 notation) or literally (`/\Q$pattern\E/` in Perl5 notation)? What the trap do you talk about?

Also, it is not clear how interpolation interacts with single and double quoting, e. g. `/'$pattern'/`, `/"$pattern"/`. Will `$pattern` be interpolated? How will it be interpreted, as regex or literally?

### PR DESCRIPTION
**Pull request by threadless-screw**
## The problem

This is a first attempt to fix #2599 and #2230

## Solution provided

Clarification of the various interpolation constructs.

### CODE REVIEW

**Comment by threadless-screw** [comment]
The reference to the [implicit EVAL trap](https://docs.perl6.org/language/traps#%3C{$x}%3E_vs_$($x):_Implicit_EVAL) does not seem very useful to me. Insofar as I understand the matter, the example provided there is rather misleading. The wording is unclear, and the failure that is presented does not result from `<{...}>` breaking, but simply from the regex being malformed: quoting the hash (#) solves the problem. I would therefore be inclined to remove that section from the docs altogether; the last comment about security issues could then be copied into the Regex interpolation section on the regexes page.

File: doc/Language/regexes.pod6 [hunk 1]
```diff
@@ -1643,62 +1650,71 @@ to emphasize the correspondence between the first two statements. Statement
...
+value.
 
 Now consider the second two syntactical forms from the table above:
 C«<$variable>» and C«<${code}>». These forms will stringify the value of the
 variable or the return value of the code and interpolate it as a regex. If the
-respective value is a Regex, it is interpolated as such. 'Interpolated as a
-regex' means interpolated/inserted into the target Regex without protective
-quoting. Consequently, the further evaluation of the target Regex may trigger
-the (second-level) interpolation of any variables it contains.   
+respective value is a Regex, it is interpolated as such:
```
Comment (JJ) [inline doc/Language/regexes.pod6]: You could use C<> here

File: doc/Language/regexes.pod6 [hunk 2]
```diff
@@ -1643,62 +1650,71 @@ to emphasize the correspondence between the first two statements. Statement
...
-Regex object into which the variable has been interpolated. In this case, C<f2>
-appends the letter 'x' to whatever string it is passed. The Regex that is
-returned by C<f2> is in turn inserted into target Regex by the C<{...}>
-construct. Statement C<[5]> illustrates another case of two-fold regex
-interpolation. When the target Regex is constructed, the strictly literal string
-value C<$x> is interpolated into it by the C<{...}> construct. When the Regex is
-evaluated further, the unprotected variable C<$x> is interpolated, i.e. replaced
-by the string value C<rege>, which explains the match.
+Importantly, 'interpolated as a regex' means interpolated/inserted into the
+target Regex without protective quoting. Consequently, if the value of the
```
Comment (JJ) [inline doc/Language/regexes.pod6]: Use simply regex or C<Regex>, depending on whether you're talking about the concept or the object.

File: doc/Language/regexes.pod6 [hunk 3]
```diff
@@ -1584,57 +1584,127 @@ documents.
 
 =head1 X<Regex interpolation|regex, Regex Interpolation>
 
-Instead of using a literal pattern for a regex match you can use a variable
-that holds that pattern.
+Instead of using a literal pattern for a regex match, you can use a variable
+that holds that pattern. This variable can then be 'interpolated' into a regex,
+such that its appearance in the regex is replaced with the pattern that it
+holds. The advantage of using interpolation this way, is that the pattern need
```
[the remark this answers is not in the document]
Comment (threadless-screw) [reply doc/Language/regexes.pod6]: This is actually correct English. 'Need' is used as an auxiliary verb here. Please see the usage note in the [dictionary](https://www.thefreedictionary.com/need).

File: doc/Language/regexes.pod6 [hunk 4]
```diff
@@ -1584,57 +1584,127 @@ documents.
...
-Here are examples showing all four ways:
+=begin table
 
-    my Str $text = 'camelia';
-    my Str $pattern0 = 'camelia';
-    my     $pattern1 = 'ailemac';
-    my str $pattern2 = '\w+';
+    Syntax         | Description
+    ===============+===========================================================
+    C«$variable»   | Evaluates stringified contents of variable literally.
```
[the remark this answers is not in the document]
Comment (threadless-screw) [reply doc/Language/regexes.pod6]: In a subsequent iteration I amended this line a bit, but as regards the issue that you point out, I think the line as it stood/stands says precisely what we both have in mind.  "Evaluating the stringified contents of a variable literally" means to first stringify the variable contents, and then interpreting the result literally. There is no ambiguity.

File: doc/Language/regexes.pod6 [hunk 5]
```diff
@@ -1584,57 +1584,127 @@ documents.
...
+illustrate a plain case of regex interpolation. Since unescaped/unquoted
+alphabetic characters in a regex match literally, the single quotes in the regex
+of statement C<[1]> are functionally redundant; they have merely been included
+to emphasize the correspondence between the first two statements. Statement
+C<[3]> unambiguously shows that the string pattern held by C<$pattern2> is
+interpreted literally, and not as a regex. In case it would have been
+interpreted as a regex, it would have matched the first word of C<$string>, i.e.
+C<｢Is｣>. Statement C<[4]> shows how the stringified number is used as a match
+pattern.
+
```
[the remark this answers is not in the document]
Comment (threadless-screw) [reply doc/Language/regexes.pod6]: I'll rearrange things a bit.

File: doc/Language/regexes.pod6 [hunk 6]
```diff
@@ -1584,57 +1584,127 @@ documents.
...
+    my sub f1          { return /s\w+/ };
+    my sub f2 (Str $x) { return /$x x/ };
+
+    say $string.match: / <$pattern1>  /;                  # [1] OUTPUT: ｢Is｣
+    say $string.match: / <$number>    /;                  # [2] OUTPUT: ｢123｣
+    say $string.match: / <{ f1 }>     /;                  # [3] OUTPUT: ｢string｣
+    my $x = "rege";
+    say $string.match: / <{ f2($x) }> /;                  # [4] OUTPUT: ｢regex｣
+
+In statement C<[4]> use if made of function C<f2>, which acts as a (very simple)
```
[the remark this answers is not in the document]
Comment (threadless-screw) [reply doc/Language/regexes.pod6]: Already fixed.

File: doc/Language/regexes.pod6 [hunk 7]
```diff
@@ -1584,57 +1584,127 @@ documents.
...
+"regex factory": you can pass it a string variable, and it will return a Regex
+object into which the variable has been interpolated. In this case, C<f2> appends
+the letter 'x' to whatever string it is passed. 
+
+Note: it may be desired to run arbitrary code from within the regex I<without>
```
[the remark this answers is not in the document]
Comment (threadless-screw) [reply doc/Language/regexes.pod6]: How about moving the full-fledged example to the other page, and leaving a brief note/link here that points the user to possible use of an ordinary/non-interpolating code block, so as to highlight the contrast between `{...}`, `$(...)` and `<$...>`?

[the remark this answers is not in the document]
Comment (JJ) [reply doc/Language/regexes.pod6]: That would be fine.

**Pull request merged**